aptos-move/framework/aptos-framework/sources/code.move

```diff
@@ -0,0 +1,185 @@
+        let packages = &mut borrow_global_mut<PackageRegistry>(addr).packages;
+        let len = vector::length(packages);
+        let index = len;
+        let i = 0;
+        while (i < len) {
+            let old = vector::borrow(packages, i);
+            if (old.name == pack.name) {
+                check_upgradability(old, &pack);
+                index = i;
+            } else {
+                check_coexistence(old, &module_names)
+            };
+            i = i + 1;
+        };
+
+        // Update registry
+        if (index < len) {
+            *vector::borrow_mut(packages, index) = pack
+        } else {
+            vector::push_back(packages, pack)
+        };
+
+        // Request publish
+        native_request_publish(addr, module_names, code, pack.upgrade_policy.policy)
+    }
+
+    /// Same as `publish_package` but as an entry function which can be called as a transaction. Because
+    /// of current restrictions for txn parameters, the metadata needs to be passed in serialized form.
+    public entry fun publish_package_txn(owner: &signer, pack_serialized: vector<u8>, code: vector<vector<u8>>)
+    acquires PackageRegistry {
+        publish_package(owner, native_from_bytes<PackageMetadata>(pack_serialized), code)
+    }
+
+    // Helpers
+    // -------
+
+    /// Checks whether the given package is upgradable, and returns true if a compatibility check is needed.
+    fun check_upgradability(old_pack: &PackageMetadata, new_pack: &PackageMetadata) {
+        assert!(old_pack.upgrade_policy.policy < upgrade_policy_immutable().policy,
+            error::invalid_argument(EUPGRADE_IMMUTABLE));
+        assert!(can_change_upgrade_policy_to( old_pack.upgrade_policy, new_pack.upgrade_policy),
+            error::invalid_argument(EUPGRADE_WEAKER_POLICY));
+    }
+
+    /// Checks whether a new package with given names can co-exist with old package.
+    fun check_coexistence(old_pack: &PackageMetadata, new_modules: &vector<String>) {
+        // The modules introduced by each package must not overlap with `names`.
+        let i = 0;
+        while (i < vector::length(&old_pack.modules)) {
+            let old_mod = vector::borrow(&old_pack.modules, i);
+            let j = 0;
+            while (j < vector::length(new_modules)) {
+                let name = vector::borrow(new_modules, j);
+                assert!(&old_mod.name != name, error::already_exists(EMODULE_NAME_CLASH))
+            }
+        }
+    }
+
+    /// Get the names of the modules in a package.
+    fun get_module_names(pack: &PackageMetadata): vector<String> {
+        let module_names = vector::empty();
+        let i = 0;
+        while (i < vector::length(&pack.modules)) {
+            vector::push_back(&mut module_names, vector::borrow(&pack.modules, i).name);
+            i = i + 1
+        };
+        module_names
+    }
+
+    /// Native function to initiate module loading
+    native fun native_request_publish(
```

Comment:
Shall we remove the "native_" prefix for the sake of consistency?

Reply:
Done